crates/pico/src/memo_ref.rs

```diff
@@ -1,13 +1,66 @@
-use std::{marker::PhantomData, ops::Deref};
+use std::{
+    any::Any,
+    hash::{Hash, Hasher},
+    marker::PhantomData,
+    ops::Deref,
+};
 
+use crate::{DatabaseDyn, DerivedNodeId, ParamId, dependency::NodeKind};
 use intern::InternId;
 
-use crate::{DatabaseDyn, DerivedNodeId, ParamId, dependency::NodeKind};
+type MemoRefProjectorStep = for<'a> fn(&'a dyn Any) -> &'a dyn Any;
+
+const MAX_PROJECTOR_STEPS: usize = 4;
+
+#[inline(always)]
+fn step_identity(value: &dyn Any) -> &dyn Any {
+    value
```

Comment:
this can panic, right? The len should always stop before we reach a step_identity, right?